### PR TITLE
Popover: fix limitShift logic by adding iframe offset correctly (and a custom shift limiter)

### PR DESCRIPTION
**Pull request by ciampo**
Tracked in #42770
Follows up to the experimentation done in #42531

## What?

This PR:
- moves the code that adds a the iframe offset from the `offset` middleware back into its own custom middleware, and
- adds a custom `limitShift` function in order that the iframe offset is taken into account correctly when shifting at the edges of the viewport.

## Why?

Currently, when the anchor is inside an `iframe` and the `Popover` has shifting enabled, the popover stops its shifting on top of its anchor instead of shifting past it.

The main issue is that the default `limitShift` function [has some custom logic when computing the min/max limits for the shifting popover](https://github.com/floating-ui/floating-ui/blob/1073602c7a42a9ad764fe6330f773c57322a1d65/packages/core/src/middleware/shift.ts#L176-L213). This logic presents some "limitations" (pun not intended) for our case:

- it adds the offset from the `offset` middleware only on the `crossAxis`
- even on the `crossAxis`, it adds that offset conditionally, based on the popover's `placement` ([min](https://github.com/floating-ui/floating-ui/blob/1073602c7a42a9ad764fe6330f773c57322a1d65/packages/core/src/middleware/shift.ts#L200), [max](https://github.com/floating-ui/floating-ui/blob/1073602c7a42a9ad764fe6330f773c57322a1d65/packages/core/src/middleware/shift.ts#L205))

I initially tried compensating for the extra iframe offset via the `offset` option for the default `limitShift` function, but that also doesn't work because that offset is added when computing the max limits, but is subtracted when computing the min limits — while, in our case, we'd _always_ need to add the iframe offset.

## How?

Based on the findings written above, the only solution is to add a custom `limiter` function for the `shift` middleware.

The new function is a modified version of [the default `limitShift` function](https://github.com/floating-ui/floating-ui/blob/1073602c7a42a9ad764fe6330f773c57322a1d65/packages/core/src/middleware/shift.ts#L147-L220). The only change is that we're always adding the extra offset for the `iframe` when computing min/max limits.

Another necessary change was to move the addition of the `iframe` offset to its own custom middleware. This has 2 advantages:

- it removes the complex logic that we had to write for the `offset` middleware
- it allows us to pass the iframe offset amount to the custom `limiter` function as a separate value from the _regular_ `offset`

## Testing Instructions

No change should be introduced for Popovers where the anchor is inside the same `document` as the popover

The easiest way to test for these changes is to observe the behavior of the block toolbar in the site editor.

Here's what I've tweaking during my tests:
- increasing the height of the toolbar on top of the iframe to test different iframe y offsets
- adding left padding to the iframe's wrapper to test different iframe x offsets
- passing an `offset` number to the popover in the block toolbar
- testing different `placement` options for the popover in the block toolbar (i.e top, bottom, left, right)

To make sure the behavior is the correct / expected one, compare the behavior of the block toolbar in the site editor (with iframe) against the behavior of the block toolbar in the post editor (no iframe).

## Screenshots or screencast 

`trunk`:

https://user-images.githubusercontent.com/1083581/186616098-06f0790a-d3e1-4297-8f9f-1bad0dd92151.mp4

This PR:

https://user-images.githubusercontent.com/1083581/186615624-52d8ab55-4582-4e6e-8479-4602fbaa2569.mp4

### CODE REVIEW

File: packages/components/src/popover/index.js
Comment (ciampo) [inline packages/components/src/popover/index.js]: We may even change this regex to force the top/bottom/left to be at the start of the string?

Comment (ntsekouras) [reply packages/components/src/popover/index.js]: Yes, a `startsWith` could be enough I think..

**Comment by ntsekouras** [comment]
>in the site editor, the toolbar keeps shifting up (when scrolling down) — while in the post editor, it stops once it goes beyond the block's top edge

I couldn't replicate this.. 🤔

File: packages/components/src/popover/index.js
Comment (ntsekouras) [inline packages/components/src/popover/index.js]: I guess we can extract a util to just return the `mainAxis and crossAxis`..

File: packages/components/src/popover/index.js
Comment (ntsekouras) [inline packages/components/src/popover/index.js]: Why do we always return the negative here?

**Review by ntsekouras — COMMENTED** [review]
This looks good Marco and I think it's really close to land(if not ready)! I've tested enough and everything seemed to work well. Just left a question 😄

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: I'm not sure if would be reducing the amount of code by much — this is because the `shift` and the `offset` middlewared use opposite logic to identify the main and cross axis. For example, if the `placement` is `top`, the main axis for the `offset` middleware is the `y` axis, while the main axis for the `shift` middleware is the `x` axis. This is just how floating ui works 🤷 

Do you still think it would be useful to abstract this logic in a separate function?

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: In my tests, setting value to a positive number never didn't lead to the behavior that we were looking for 🤷  Although it would be great if @talldan (who also has experience working and debugging these middlewares) could confirm that this is the correct "math"

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: I had initially used a RegEx for its compact form, but I switched to `startsWith` as you suggested in https://github.com/WordPress/gutenberg/pull/42950/commits/43c1d8db0013d15e85c51e8ceeda7132639e06e7

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: Instinctively, I would say that the we change sign of the `crossAxis` to take into account the same change of sign that the `offset` middleware applies to its `mainAxis`. Keeping in mind that the two middlewares have opposite definitions for cross and main axis, this means that the change of sign always applied to the axis (always `x` or always `y`)

**Comment by ciampo** [comment]
> > in the site editor, the toolbar keeps shifting up (when scrolling down) — while in the post editor, it stops once it goes beyond the block's top edge
> 
> I couldn't replicate this.. 🤔

Here is what I see on my branch after applying this change

```diff
diff --git a/packages/block-editor/src/components/block-popover/index.js b/packages/block-editor/src/components/block-popover/index.js
index 370959d1b1..8aad54a2d8 100644
--- a/packages/block-editor/src/components/block-popover/index.js
+++ b/packages/block-editor/src/components/block-popover/index.js
@@ -53,7 +53,7 @@ export default function BlockPopover( {
 		<Popover
 			ref={ popoverScrollRef }
 			animate={ false }
-			position="top right left"
+			placement="left"
 			focusOnMount={ false }
 			anchorRef={ anchorRef }
 			// Render in the old slot if needed for backward compatibility,
```

In the site editor (iframed), you can notice that the popover keeps shifting when scrolling UP


https://user-images.githubusercontent.com/1083581/183721836-fed88ad3-2bf2-48ad-82ea-2ae0bb280646.mp4


Here you can see the behavior of the same popover in the post editor (without iframe):


https://user-images.githubusercontent.com/1083581/183722198-e4c30922-afd8-44d8-9583-fc3af25d63f4.mp4

Interestingly, this behavior is probably related to the `limitShift` value passed for the `mainAxis` (also discussed [here](https://github.com/WordPress/gutenberg/pull/42950#discussion_r941311798))

File: packages/components/src/popover/index.js
Comment (ntsekouras) [reply packages/components/src/popover/index.js]: Oh sorry, I missed that.. It's good as is.

File: packages/components/src/popover/index.js
Comment (ntsekouras) [reply packages/components/src/popover/index.js]: Yes, it seems the offset plays its part here and the `left` behavior issue is exactly 2 times the iframe offset.. I'm not sure yet if `shift` uses the `offset` value from the middleware and for other calculation takes into account the `iframe` so its doubled..

**Comment by ciampo** [comment]
Update: I believe the unexpected behavior that I flagged above could be related to a limitation of the `floating-ui` library — I opened https://github.com/floating-ui/floating-ui/issues/1859 to hopefully get more insights (and potentially work on a fix)

File: packages/components/src/popover/index.js
Comment (ciampo) [reply packages/components/src/popover/index.js]: Looking at the source code, it looks like the offset from the `offset` middleware doesn't get taken into account for the `mainAxis` (while it does get taken into account for the `crossAxis`).

This discrepancy may be the root cause for the unexpected behavior that we've been experiencing on this PR. I've opened https://github.com/floating-ui/floating-ui/issues/1859, let's see how that goes

**Comment by ciampo** [comment]
I did some more testing, and I believe that the only way to get the to the behaviour that we want (while generalizing it enough to work for different `placements`) is to modify the `limitShift` function itself.

I tried that in https://github.com/WordPress/gutenberg/pull/42950/commits/5a1bc5e75327057f7b5152fcb488b0115392c854, and it seems promising — even if I'd really want to avoid implementing custom middleware like that. That's why I opened [a PR on the `@floating-ui` repo](https://github.com/floating-ui/floating-ui/pull/1868) which applies the same changes the default `limitShift` function — I'm not sure it's going to be approved, but maybe it's going to give me some more insights into how it's supposed to work.

**Comment by ciampo** [comment]
@ntsekouras and @talldan , I believe this is ready for another round of review.

I've updated the PR description too, for more details on the solution i'm proposing.

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: Just noting that I'll have to move this entry to the unreleased section above

**Comment by ntsekouras** [comment]
Thanks for your work Marco!

I'm a bit hesitant to have some kind of 'fork' of limitShift function(also this [comment](https://github.com/floating-ui/floating-ui/pull/1868#issuecomment-1226956006)), but not sure if we can avoid something like this eventually..

>but that also doesn't work because that offset is added when computing the max limits, but is subtracted when computing the min limits — while, in our case, we'd always need to add the iframe offset.

Since we know the problem, do you think we could explore something simpler using math like this: 
https://github.com/WordPress/gutenberg/pull/43162/files#diff-afbc365e3b220c93528fb5ced625536dbcaa8cfcd0858aa028908ab22d6d0488R312? It might need some adjustments but overall seemed to work well. 🤔

**Comment by ciampo** [comment]
I personally prefer the current approach in this PR more, because:

 - the way we are applying the frame offset in the `offset` middleware was hacky anyway (we were already countering the logic of the `limitShift` function, as explained [in the PR description](https://github.com/WordPress/gutenberg/pull/42417#issue-1304501346)), and therefore prone to break if the `limitShift` function's logic changed
 - the [approach that you propose](https://github.com/WordPress/gutenberg/pull/43162/files#diff-afbc365e3b220c93528fb5ced625536dbcaa8cfcd0858aa028908ab22d6d0488R312) still feels as hacky to me:
   - we would be trying to counter the internal logic of `limitShift` — which means that, if the `limitShift` logic changes internally, it could also still break this approach
   - from what I gathered during my time spent on this issue, that logic can't be really "fixed" from outside the function from a math standpoint
   - I haven't tested it thoroughly, but I'm not sure that it can be mathematically correct for all scenarios (all the placements, negative `offset`, etc...)


Also, I'm discussing with the author of the library the [possibility of adding a new flag to their `limitShift` function](https://github.com/floating-ui/floating-ui/pull/1868#issuecomment-1228205809), which could allow to always add the frame offset to min/max limits (which means that we could remove our fork of `limitShift`) and therefore we would be able, in the future, to remove our custom implementation and use again the library's one.

**Comment by ntsekouras** [comment]
>Also, I'm discussing with the author of the library the https://github.com/floating-ui/floating-ui/pull/1868#issuecomment-1228205809, which could allow to always add the frame offset to min/max limits

Do you think this is something that could land quite soon and we could wait?

**Comment by ciampo** [comment]
> Do you think this is something that could land quite soon and we could wait?

Difficult to tell at the moment, it will mostly depend on what the library author (@atomiks) thinks

**Comment by ciampo** [comment]
> > Also, I'm discussing with the author of the library the [floating-ui/floating-ui#1868 (comment)](https://github.com/floating-ui/floating-ui/pull/1868#issuecomment-1228205809), which could allow to always add the frame offset to min/max limits
> 
> Do you think this is something that could land quite soon and we could wait?

Update — I've opened [a new issue](https://github.com/floating-ui/floating-ui/issues/1870) to track that conversation 

---

Apart from that, I still think that regardless of https://github.com/floating-ui/floating-ui/issues/1870, this PR implements a better solution than the current one to handle limit shifting.

File: packages/components/src/popover/limit-shift.ts
Comment (talldan) [inline packages/components/src/popover/limit-shift.ts:64]: I'm no TS expert, but I wonder if it's worth extracting this to a separate type to make this easier to read ...

e.g.
```ts
type LimitShiftOffset =
	| ( ( args: MiddlewareArguments ) =>
			| number
			| axisOptions ) 
	| number 
	| axisOptions 
```

Although I'm not sure if this is a case where you want to keep the code as close to the floating-ui implementation as possible.

**Review by talldan — APPROVED** [review]
This tests well for me and fixes the issue.

The only question I have is whether the bug experienced in #42531 (with left/right placements in the site editor frame) is critical enough to warrant the added complexity of this fix. It might be ok just to live with that issue until there's a floating-ui fix. I'm not sure Gutenberg has any left/right placements that would be affected by it, and I think it's a niche enough bug that other consumers of the components package are unlikely to experience it. Even if they do, it isn't the worst bug in the world.

The risk isn't so much the code as it stands now, but that over time more fixes are added to `customLimitShift`, and it becomes out of sync with how floating-ui's shift works.

I'll leave it up to you.

**Comment by ciampo** [comment]
Thank you for your review, Dan!

> I'm not sure Gutenberg has any left/right placements that would be affected by it, and I think it's a niche enough bug that other consumers of the components package are unlikely to experience it. Even if they do, it isn't the worst bug in the world.

I would say that the most important bug that this PR would fix is that currently the block toolbar in the site editor stops scrolling on top of a block, instead of under it (see PR description), but I agree that it's not a major bug like others we've been fixing.

> The risk isn't so much the code as it stands now, but that over time more fixes are added to customLimitShift, and it becomes out of sync with how floating-ui's shift works.

I agree with your point! The point I was trying to make is that even the code that we have currently (adding the iframe offset with a different sign, depending on the axis and the placement) has a similar potential to break, if the `limitShift` logic changes — and so, my argument was that thing change could be a better trade-off.

---

I think the best thing to say here is to NOT merge this PR for the time being. I will try to focus my efforts in trying to improve cross-document support directly in `floating-ui` (https://github.com/floating-ui/floating-ui/issues/1869):

- if things go well, hopefully we'll be able to rewrite our `Popover` removing a lot of the patches that we've added to support the `iFrame` (including the custom `offset` middleware code)
- if things don't go as well (or as fast), we can still re-consider merging this PR before the release of WordPress 6.1

What do y'all think?

**Comment by ciampo** [comment]
Actually, as I was playing around with the editor testing for other fixes, I found another bug that shows [the limits](https://github.com/WordPress/gutenberg/pull/42950#issuecomment-1228254542) of our current approach.

On `trunk`:

https://user-images.githubusercontent.com/1083581/187265502-bd246a28-a239-4454-98ef-27407fecae3b.mp4

Basically, the toolbar is rendered outside of the viewport — when I debugged the problem initially, I realized that the problem was only appearing when the `shift` middleware was active. I looked a bit more into it, and found out that problem is caused by the current custom logic that we have in the `offset` middleware, which is then not compatible is such edge cases with the `shift` and `limitShift` logic.

I then applied code changes from this PR, and it seems that the bug is solved:

https://user-images.githubusercontent.com/1083581/187266288-da94daed-16ba-4a26-a6d8-fde91baf1e4d.mp4

**Comment by ciampo** [comment]
Update:

I've started working on cross-document support directly in `floating-ui` (https://github.com/floating-ui/floating-ui/pull/1884), but I don't think I'll manage to wrap that up in time for 6.1.

My suggestion is that we merge this PR as a temporary fix, since it solves a few issues related to shift limiting. We will be able to remove the temporary fix once the `floating-ui` improvements will be released.

What do you think, @ntsekouras  and @talldan  ?

File: packages/components/src/popover/limit-shift.ts
Comment (ntsekouras) [inline packages/components/src/popover/limit-shift.ts:40]: We might need some more comments [like here](https://github.com/WordPress/gutenberg/pull/44112/files#diff-b1eae6bea5fe8fa3c1b81d05c659aa377dd3f7c6f90fd8ad3a41e0ec178c0afdR1)

**Review by ntsekouras — COMMENTED** [review]
Thanks for all your work and for the PR in the floating-ui. Improving the library for cross-document support is definitely going to take time, so I guess you can rebase and we can do the final testing.

File: packages/components/src/popover/limit-shift.ts
Comment (ciampo) [reply packages/components/src/popover/limit-shift.ts:64]: > I'm no TS expert, but I wonder if it's worth extracting this to a separate type to make this easier to read

I definitely agree that parts of this type could be extracted to make the overall type less verbose / easier to understand

> Although I'm not sure if this is a case where you want to keep the code as close to the floating-ui implementation as possible

That is the rationale that I adopted when writing the custom limiter function. I basically tried to make the least amount of changes possible compared to the original function.

For now I've kept the code as-is, but let me know if you have a strong preference for simplifying the types and I can definitely do something about it.

File: packages/components/src/popover/limit-shift.ts
Comment (ciampo) [reply packages/components/src/popover/limit-shift.ts:40]: Good shout! Done in https://github.com/WordPress/gutenberg/pull/42950/commits/c2925087bebf3fbec0e56c304d6083cea7f9b247

**Comment by ciampo** [comment]
> I guess you can rebase and we can do the final testing.

I just rebased and pushed the updated changes — The PR should be ready for a final review 🙏

**Review by ntsekouras — APPROVED** [review]
Thanks Marco!

**Comment by ciampo** [comment]
@c4rl0sbr4v0 @ockham could you please include this PR in WP 6.1 beta 2? Thank you 🙏

**Comment by ockham** [comment]
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 896fc77e4a

**Pull request merged**
